ui/src/app/views/workspace-share/component.tsx

```diff
@@ -0,0 +1,405 @@
+    marginTop: '1rem',
+    letterSpacing: '.01em',
+    color: '#565656'
+  },
+
+  dropdown: {
+    width: '100%',
+    marginTop: '0',
+    borderBottom: '1px solid gray',
+    paddingBottom: '.5rem',
+  },
+
+  open: {
+    overflow: 'hidden',
+    position: 'absolute',
+    backgroundColor: '#FFFFFF',
+    border: '1px solid'
+  },
+
+  noBorder: {
+    border: 'none',
+    background: 'none',
+    width: '90%'
+  },
+
+  spinner: {
+    animation: '1s linear infinite spin',
+    position: 'relative',
+    display: 'inline-block',
+    verticalAlign: 'text-bottom'
+  },
+
+  dropdownMenu: {
+    display: 'block',
+    maxHeight: '12rem',
+    minHeight: '30px',
+    visibility: 'visible',
+    overflowY: 'scroll',
+    maxWidth: '16rem',
+    width: '100%',
+    marginTop: '.25rem'
+  },
+
+  wrapper: {
+    display: 'grid',
+    gridTemplateColumns: '9rem 0.1rem 0.1rem',
+    gridGap: '10px',
+    alignItems: 'center',
+    height: '6.38%'
+  },
+
+  box: {
+    borderRadius: '5px',
+    paddingTop: '0.2rem',
+    paddingLeft: '0.2rem'
+  },
+
+  userName: {
+    marginTop: '0',
+    paddingTop: '0',
+    fontSize: '0.58333rem',
+    height: '4.5%',
+    color: '#888',
+    fontFamily: 'Montserrat',
+    lineHeight: '1rem',
+    whiteSpace: 'nowrap'
+  },
+
+  collaboratorIcon: {
+    margin: '0 0 0 4rem',
+    color: '#2691D0',
+    cursor: 'pointer'
+  },
+
+  sharingBody: {
+    maxHeight: '70vh',
+    overflowY: 'hidden',
+    overflowX: 'visible',
+    padding: '0 .125rem',
+    display: 'flex',
+    flexDirection: 'column'
+  }
+
+});
+
+export interface WorkspaceShareState {
+  autocompleteLoading: boolean;
+  autocompleteUsers: User[];
+  userNotFound: string;
+  workspaceShareError: boolean;
+  usersLoading: boolean;
+  userRolesList: UserRole[];
+  workspaceFound: boolean;
+  workspaceUpdateConflictError: boolean;
+  workspace: Workspace;
+  searchTerm: string;
+}
+
+export interface WorkspaceShareProps {
+  workspace: Workspace;
+  accessLevel: WorkspaceAccessLevel;
+  userEmail: string;
+  closeFunction: Function;
+  sharing: boolean;
+}
+
+export class WorkspaceShare extends React.Component<WorkspaceShareProps, WorkspaceShareState> {
+  searchTermChangedEvent: Function;
+
+  constructor(props: WorkspaceShareProps) {
+    super(props);
+    this.state = {
+      autocompleteLoading: false,
+      autocompleteUsers: [],
+      userNotFound: '',
+      workspaceShareError: false,
+      usersLoading: false,
+      userRolesList: this.props.workspace.userRoles || [],
+      workspaceFound: (this.props.workspace !== null),
+      workspaceUpdateConflictError: false,
+      workspace: this.props.workspace,
+      searchTerm: '',
+    };
+    this.searchTermChangedEvent = fp.debounce(300, this.userSearch);
+  }
+
+  save(): void {
+    if (this.state.usersLoading) {
+      return;
+    }
+    this.setState({usersLoading: true, workspaceShareError: false});
+    workspacesApi().shareWorkspace(this.state.workspace.namespace,
```

Comment:
The stub doesn't have an implementation for `shraeWorkspace()` - so how are the tests working? Am I missing something?

Reply:
Oooh. Good point, I actually test for all the functionality of the modal (reactive to user additions/removals, etc) but not the save function! I'll add that now.

Reply:
tests added